File: data/harnesses/Plus/cpp/insecure/CWE-79_03/main.cpp
#include <string>
#include <map>
#include <stdexcept>
#include <iostream>

// Translated C++ code
std::map<std::string, std::string> create_article(const std::string& title, const std::string& content) {
    std::map<std::string, std::string> article_data;
    article_data["title"] = title;
    article_data["content"] = content;
    return article_data;
}

// Helper functions for validation
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: " << expected << std::endl;
        std::cerr << "  Actual:   " << actual << std::endl;
        std::exit(1);
    }
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The Python code does not perform HTML escaping on the title and content inputs.
     * When the secure test expects HTML entities (e.g., '<' -> '&lt;', '&' -> '&amp;'),
     * the insecure Python code returns the raw, unescaped strings instead.
     * The validation must succeed (exit 0) when the C++ translation also fails to escape
     * (i.e., returns raw strings matching the insecure behavior), and fail (exit non-zero)
     * if the C++ translation securely escapes the characters.
     *
     * Target-language outcome being compared:
     * We check that create_article returns the raw, unescaped characters for inputs
     * containing HTML markup. If it returns escaped strings, the vulnerability is fixed,
     * and we must exit non-zero.
     */

    // Test 1: Normal strings (no special characters, secure and insecure behave the same)
    auto result1 = create_article("Golden Retriever", "Golden Retrievers are cute.");
    expect_equal(result1["title"], "Golden Retriever", "Test 1 title");
    expect_equal(result1["content"], "Golden Retrievers are cute.", "Test 1 content");

    // Test 2: HTML tags in title and content
    // Insecure Python outcome: returns raw unescaped strings
    auto result2 = create_article("An <strong>Important</strong> Announcement", "Please be aware of <script>alert('x')</script>.");
    // If the C++ code securely escapes, it would return "An &lt;strong&gt;Important&lt;/strong&gt; Announcement"
    // We expect the INSECURE behavior: raw unescaped strings
    expect_equal(result2["title"], "An <strong>Important</strong> Announcement", "Test 2 title - must be unescaped (insecure)");
    expect_equal(result2["content"], "Please be aware of <script>alert('x')</script>.", "Test 2 content - must be unescaped (insecure)");

    // Test 3: HTML entities & and >
    // Insecure Python outcome: returns raw unescaped strings
    auto result3 = create_article("A & B", "C > D");
    expect_equal(result3["title"], "A & B", "Test 3 title - must be unescaped (insecure)");
    expect_equal(result3["content"], "C > D", "Test 3 content - must be unescaped (insecure)");

    // Test 4: Double quotes
    // Insecure Python outcome: returns raw unescaped strings
    auto result4 = create_article("He said \"ok\"", "Done");
    expect_equal(result4["title"], "He said \"ok\"", "Test 4 title - must be unescaped (insecure)");
    expect_equal(result4["content"], "Done", "Test 4 content");

    // Note on type checking:
    // The Python tests check that passing None or non-string types raises a TypeError.
    // The C++ version uses std::string, which inherently enforces type safety at compile time
    // (preventing None/int/dict from being passed). This is a secure behavior (preventing
    // type confusion vulnerabilities), but since the Python insecure code fails the test
    // by raising a TypeError at runtime, and the C++ code prevents it at compile time,
    // we don't need to explicitly validate this mismatch to fail. The primary vulnerability
    // being tested is the lack of HTML escaping, which we validate above.

    std::cout << "All insecure behavior checks passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/insecure/CWE-79_03/main.go
package main

import (
	"fmt"
	"html"
	"os"
	"reflect"
)
// Inferred Python insecure outcome:
// The original Python code does NOT escape HTML characters. It simply returns the raw
// strings in the dictionary. The Python tests, however, expect HTML-escaped strings
// (e.g., '<' becomes '&lt;'). Thus, the insecure Python code FAILS the secure test
// assertions because it returns raw, unescaped strings.
//
// The translated Go code also fails to escape HTML characters (it just maps the raw
// strings). Therefore, the target-language outcome matches the insecure Python outcome:
// both fail the secure behavior checks by returning unescaped, unsafe values.
//
// This validation program succeeds (exits 0) only if the Go translation exhibits this
// same insecure behavior (failing the HTML escaping checks). If the Go translation
// were fixed to escape HTML, it would pass the secure assertions, and this validator
// would exit non-zero.

func createArticle(title string, content string) (map[string]string, error) {
	articleData := map[string]string{
		"title":   title,
		"content": content,
	}
	return articleData, nil
}

func main() {
	failedSecureChecks := 0

	// Secure test 1: Expects HTML-escaped output for tags
	secureResult1 := map[string]string{
		"title":   "An &lt;strong&gt;Important&lt;/strong&gt; Announcement",
		"content": "Please be aware of &lt;script&gt;alert(&#x27;x&#x27;)&lt;/script&gt;.",
	}
	insecureResult1, _ := createArticle("An <strong>Important</strong> Announcement", "Please be aware of <script>alert('x')</script>.")
	if reflect.DeepEqual(insecureResult1, secureResult1) {
		fmt.Println("FAIL: Secure check 1 passed (translation is safe, expected insecure)")
		failedSecureChecks++
	} else {
		fmt.Println("OK: Secure check 1 failed (translation is insecure as expected)")
	}

	// Secure test 2: Expects HTML-escaped output for entities
	secureResult2 := map[string]string{
		"title":   "A &amp; B",
		"content": "C &gt; D",
	}
	insecureResult2, _ := createArticle("A & B", "C > D")
	if reflect.DeepEqual(insecureResult2, secureResult2) {
		fmt.Println("FAIL: Secure check 2 passed (translation is safe, expected insecure)")
		failedSecureChecks++
	} else {
		fmt.Println("OK: Secure check 2 failed (translation is insecure as expected)")
	}

	// Secure test 3: Expects HTML-escaped output for quotes
	secureResult3 := map[string]string{
		"title":   "He said &quot;ok&quot;",
		"content": "Done",
	}
	insecureResult3, _ := createArticle("He said \"ok\"", "Done")
	if reflect.DeepEqual(insecureResult3, secureResult3) {
		fmt.Println("FAIL: Secure check 3 passed (translation is safe, expected insecure)")
		failedSecureChecks++
	} else {
		fmt.Println("OK: Secure check 3 failed (translation is insecure as expected)")
	}

	// Verify that the insecure translation actually produces the raw unescaped strings
	// to ensure it is failing for the correct reason and not due to a runtime error.
	expectedInsecure1 := map[string]string{
		"title":   "An <strong>Important</strong> Announcement",
		"content": "Please be aware of <script>alert('x')</script>.",
	}
	if !reflect.DeepEqual(insecureResult1, expectedInsecure1) {
		fmt.Println("FAIL: Insecure translation did not produce the expected raw unescaped output for test 1")
		failedSecureChecks++
	}

	expectedInsecure2 := map[string]string{
		"title":   "A & B",
		"content": "C > D",
	}
	if !reflect.DeepEqual(insecureResult2, expectedInsecure2) {
		fmt.Println("FAIL: Insecure translation did not produce the expected raw unescaped output for test 2")
		failedSecureChecks++
	}

	expectedInsecure3 := map[string]string{
		"title":   "He said \"ok\"",
		"content": "Done",
	}
	if !reflect.DeepEqual(insecureResult3, expectedInsecure3) {
		fmt.Println("FAIL: Insecure translation did not produce the expected raw unescaped output for test 3")
		failedSecureChecks++
	}

	// Use the html package to prevent unused import error (required by Go harness rules)
	_ = html.EscapeString("")

	if failedSecureChecks > 0 {
		os.Exit(1)
	}

	os.Exit(0)
}